### ISSUE TITLE
Aw snap, my EV3 LCD isn't working!

### ISSUE THREAD

**Issue by luRaichu** [ev3dev/ev3dev#1603]
Today I cleaned my EV3's LCD with compressed air (cause it was full of dust) and after re-plugging the FPC cable (and trying again multiple times) the screen wouldn't turn on. Then, to top it all off I accidentally put a minor crack in the bottom-right corner of the LCD. I guess I need a new one!
[This tutorial](https://youtu.be/gPNJC5Uz9HY) requires surface-mount soldering an FPC socket to a breakout board. I can't do surface-mount with pins that tiny. Is there a way to replace the LCD with only through-hole soldering?

**Comment by luRaichu** [comment]
I just found a 16-pin FPC breakout with everything presoldered: https://tinkersphere.com/cables-wires/3643-16-pin-05mm-1mm-pitch-fpc-to-dip-breakout.html

**Comment by luRaichu** [comment]
I already ordered an FPC cable, breakout and color screen. I'm currently waiting on the breakout but would it be possible to dim/turn off the backlight on the new LCD?

**Comment by dlech** [comment]
The devicetree just has it configured as a gpio backlight, so it is just on/off.

**Comment by luRaichu** [comment]
Just tried to hook up the Fermion 1.8" 128x160 TFT I got tonight, I'm getting no sign of life. I'll try again tomorrow.
Is there anything I can measure to make sure the EV3 is outputting proper voltages to the FPC connector? And I'd certainly like to know how to make sure my TFT isn't dead.

**Comment by luRaichu** [comment]
![IMG_4870](https://github.com/ev3dev/ev3dev/assets/69605426/90da04ca-85e5-4e17-b4b1-1c6fd1033cb8)
I redid the wiring but the screen is still blank. And yes, I checked the continuity. Everything's connected and kosher.
I might have possibly killed my TFT in the process of removing unnecessary interfaces.

**Comment by dlech** [comment]
It could also just be that the backlight is turned off. So the first thing I would measure is the voltage on the backlight pin. That should be independent of the rest of the display. If you have voltage there and on the VCC pin and the display isn't lighting up, then yeah, you might have broken something.

If there isn't voltage on the backlight pin, then I would SSH into the EV3 to make sure the correct device tree is being used.

**Comment by luRaichu** [comment]
There's 3.3v on the backlight pin. The screen must be dead.

**Comment by luRaichu** [comment]
I just received a brand new replacement TFT, I wired it up in [the same fashion](https://github.com/ev3dev/ev3dev/issues/938#issuecomment-1576648957) but it still won't light up. There's 3.3v on the backlight pin, but for a moment I accidentally wired CSDIS (EV3) -> M_CS_SD (TFT) and MISODIS  (EV3)-> BLK (TFT). I powered on the EV3 with the screen attached but then shut it off after I noticed this error, corrected it, and it still won't light up. Did the temporary wiring error break something?

There's no video signal on-screen either, so it's not just the backlight being off.

**Comment by dlech** [comment]
Is there voltage on the VCC pin? Are you sure the EV3 is using the correct device tree file?

**Comment by luRaichu** [comment]
Measuring VCC against GND, there's only half a volt flowing through. As for the device tree I'm not sure how to evaluate and correct it.

**Comment by dlech** [comment]
You should be able to SSH into the EV3 and get a terminal so you can inspect the kernel logs/sysfs to see what is loaded.

But VCC should be 3.3V and be always on, so maybe there is a problem with the wiring?

**Comment by luRaichu** [comment]
Turns out the wiring was backwards. When using the Tinkersphere FPC breakout for this project you must follow the pin numbering from the opposite side.
Now, the TFT lights up but then switches to a static pattern. Am I supposed to enable the Adafruit driver in uEnv.txt?

**Comment by dlech** [comment]
> Am I supposed to enable the Adafruit driver in uEnv.txt?

yes

```
# uncomment this line if you have an Adafruit 1.8" screen
fdtfile=da850-lego-ev3-adafruit18.dtb
```

**Comment by luRaichu** [comment]
Alright, it's finally working now, however for my mod I want the screen rotated 180 degrees from what it is now. I'll also try to compile the ev3dev kernel and correct the 1 column and 2 rows that are glitched out.
![IMG_5343_CROP](https://github.com/ev3dev/ev3dev/assets/69605426/378e79af-e4a4-486b-95e1-1513e70d634b)

**Comment by dlech** [comment]
Finally! That's really great!

FYI, you can use the `da850-lego-ev3-color.dtb` instead of the adafruit18 one and it should take care of the rotation. But yeah, will probably have to modify the kernel driver to take care of the bad pixels on the edges.

**Comment by luRaichu** [comment]
Ooh, I'm booting right now with the `da850-lego-ev3-color.dtb` device driver and the bad pixels are gone...

**Comment by luRaichu** [comment]
![IMG_5344_CROPls](https://github.com/ev3dev/ev3dev/assets/69605426/24412422-5fba-429e-98ce-1c31ca3a3916)
With `da850-lego-ev3-color.dtb` the screen is still offset though. What do I add to the st7735r.c driver to correct it?

**Comment by luRaichu** [comment]
I'm attempting to compile the ev3dev-stretch kernel on Xubuntu 23.10 (even though I haven't updated the st7735r driver yet) and I'm hit with this make error:
```
make[1]: Leaving directory '/home/bigboys/kernelbuild/ev3/ev3dev-buildscripts/build-area/linux-ev3dev-ev3-obj'
make: *** [Makefile:146: sub-make] Error 2
make: Leaving directory '/home/bigboys/kernelbuild/ev3/ev3-kernel'
```
I'm following [this tutorial](https://github.com/ev3dev/ev3dev-buildscripts/blob/ev3dev-stretch/README.md).

**Comment by dlech** [comment]
Is this the full output? I would expect the actual error to be listed before this.

**Comment by luRaichu** [comment]
Here's the full build log in all its glory:
https://pastebin.com/3EpqyzHx
Of note is this
```
/home/bigboys/kernelbuild/ev3/ev3-kernel/arch/arm/mm/proc-arm926.S: Assembler messages:
/home/bigboys/kernelbuild/ev3/ev3-kernel/arch/arm/mm/proc-arm926.S:477: Error: junk at end of line, first unrecognized character is `#'
make[2]: *** [/home/bigboys/kernelbuild/ev3/ev3-kernel/scripts/Makefile.build:427: arch/arm/mm/proc-arm926.o] Error 1
make[1]: *** [/home/bigboys/kernelbuild/ev3/ev3-kernel/Makefile:1038: arch/arm/mm] Error 2
```

**Comment by dlech** [comment]
It sounds like the wrong assembler program is being used (one that doesn't support GCC extensions).

**Comment by luRaichu** [comment]
I successfully compiled the x86 Linux kernel (6.7) earlier...

**Comment by dlech** [comment]
Maybe `ARCH=arm` or `CROSS_COMPILE=arm-linux-gnueabi-` is missing? Or there is some leftover generated file from the previous build interfering?

**Comment by dlech** [comment]
> What do I add to the st7735r.c driver to correct it?

I thinks something like this: https://github.com/ev3dev/ev3-kernel/blob/ad1f6c8727bce29ec400c9385f7f72ee2c0e59fc/drivers/gpu/drm/tinydrm/st7586.c#L221

But with non-zero starting addresses.

**Comment by luRaichu** [comment]
How do I turn off the backlight?

**Comment by dlech** [comment]
One way it to poke it directly via `/sys/class/backlight/`.

**Comment by luRaichu** [comment]
Backlight brightness lives at `/sys/class/backlight/backlight/brightness`, but you can only turn it on/off if you have the backlight pin connected. You can switch it like this:
```
echo 0 | sudo tee /sys/class/backlight/backlight/brightness
```
I'd like to fix it so you can do this without root. And perhaps you could simply use PWM for dimming?
As for those pesky glitched pixels, I built a custom kernel on a Ubuntu LTS VPS with this mod to the st7586 C driver like you said:
```
mipi_dbi_command(mipi, MIPI_DCS_SET_PARTIAL_AREA, 0x01, 0x02, 0x03, 0x77);
```
This has no effect, probably because the Fermion and Adafruit TFTs actually use the st7735r driver instead. However, there are some clues in `/include/video/mipi_display.h`:
```h
/* MIPI DSI Processor-to-Peripheral transaction types */
enum {
	MIPI_DSI_V_SYNC_START				= 0x01,
	MIPI_DSI_V_SYNC_END				= 0x11,
	MIPI_DSI_H_SYNC_START				= 0x21,
	MIPI_DSI_H_SYNC_END				= 0x31,
```
I'll play with these and report if they do anything.

**Comment by luRaichu** [comment]
> ```c
> /* MIPI DSI Processor-to-Peripheral transaction types */
> enum {
> 	MIPI_DSI_V_SYNC_START				= 0x01,
> 	MIPI_DSI_V_SYNC_END				= 0x11,
> 	MIPI_DSI_H_SYNC_START				= 0x21,
> 	MIPI_DSI_H_SYNC_END				= 0x31,
> ```
> 
> I'll play with these and report if they do anything.

These do nothing.

**Comment by dlech** [comment]
> And perhaps you could simply use PWM for dimming?

Possible, yes. Simple, no. 

The pin it is connected to doesn't have a hardware PWM that can be muxed. There is a bitbang gpio-pwm driver, but that would cause the screen to flicker because it isn't a hardware PWM with perfect timing. So the only way I know to make it work nicely would be add another PWM to the PRU firmware like we did for the status lights. https://github.com/ev3dev/ev3dev-pru-firmware/blob/master/ev3-pru1/main.c

> This has no effect, probably because the Fermion and Adafruit TFTs actually use the st7735r driver instead.

I meant that the line should be copied and pasted to the st7735r driver. It uses (mostly) the same command set. There may be an additional command required to enable it too (`MIPI_DCS_ENTER_PARTIAL_MODE`?). The datasheet should hopefully explain this (note: the datasheet calls command 0x30 `PTLAR` but is the same value as `MIPI_DCS_SET_PARTIAL_AREA`)

**Comment by dlech** [comment]
> I'd like to fix it so you can do this without root.

Normally, the display manager would do this. But our display manager (brickman) was designed mainly for the default display that doesn't have a backlight so we never included this.

Perhaps the simplest thing to do would be to add a udev rule similar to `/lib/udev/rules.d/60-ev3dev.rules` to allow users in the `ev3dev` group access to the file in sysfs. (Don't modify this file, create a new on in `/etc/udev/rules.d/`.) Pro tip: don't forget the newline at the end of the file - IIRC this version of udev will ignore the last line of the file without it.

```
SUBSYSTEM=="backlight", ACTION=="add", \
    RUN+="/usr/bin/find /sys$devpath -type f -exec /bin/chmod g+u {} + -exec /bin/chown :ev3dev {} +"
```

Or you could make a daemon program that runs in the background  as root and uses some sort of IPC to allow users control the backlight.

**Comment by luRaichu** [comment]
> Perhaps the simplest thing to do would be to add a udev rule similar to `/lib/udev/rules.d/60-ev3dev.rules` to allow users in the `ev3dev` group access to the file in sysfs. (Don't modify this file, create a new on in `/etc/udev/rules.d/`.) Pro tip: don't forget the newline at the end of the file - IIRC this version of udev will ignore the last line of the file without it.
> 
> ```
> SUBSYSTEM=="backlight", ACTION=="add", \
>     RUN+="/usr/bin/find /sys$devpath -type f -exec /bin/chmod g+u {} + -exec /bin/chown :ev3dev {} +"
> ```

That works. As for the st7735r driver, I added
```
mipi_dbi_command(mipi, MIPI_DCS_SET_PARTIAL_AREA, 0x00, 0x00, 0x00, 0x77);
```
Then I changed `MIPI_DCS_ENTER_NORMAL_MODE` to `MIPI_DCS_ENTER_PARTIAL_MODE` and this is the result
![IMG_5392_CROP](https://github.com/ev3dev/ev3dev/assets/69605426/068437aa-e81e-4b58-8e28-9e6cfe9e0a3b)
There are still bad pixels on the bottom of the screen (which are not visible in the photo). I don't think the `MIPI_DCS_ENTER_PARTIAL_MODE` command can change the actual offset of the display?

**Comment by dlech** [comment]
Maybe that is the wrong command. Looking deeper, think we need to apply this patch series. https://patchwork.freedesktop.org/series/71567/ Then there will be a way to add the required offsets needed.

**Comment by luRaichu** [comment]
Yeah, the ev3dev kernel is too old for this patch to directly apply. You'd have to backport it manually, which is beyond my expertise.

**Comment by dlech** [comment]
I think the key patch is this one. https://patchwork.freedesktop.org/patch/348843/?series=71567&rev=3

You could search for `MIPI_DCS_SET_COLUMN_ADDRESS` and `MIPI_DCS_SET_PAGE_ADDRESS` and add the offsets there, just hard-coded instead of adding the `left_offset` and `top_offset` to the struct.

**Comment by luRaichu** [comment]
Is there any documentation on the standard use/parameters of these MIPI commands? That would help me figure out a couple things on my own.

**Comment by luRaichu** [comment]
The `SET_COLUMN` and `SET_PAGE` commands live in `drm_mipi_dsi.c`, but I'm not sure what to modify here.
```c
/**
 * mipi_dsi_dcs_set_column_address() - define the column extent of the frame
 *    memory accessed by the host processor
 * @dsi: DSI peripheral device
 * @start: first column of frame memory
 * @end: last column of frame memory
 *
 * Return: 0 on success or a negative error code on failure.
 */
int mipi_dsi_dcs_set_column_address(struct mipi_dsi_device *dsi, u16 start,
                                    u16 end)
{
        u8 payload[4] = { start >> 8, start & 0xff, end >> 8, end & 0xff };
        ssize_t err;

        err = mipi_dsi_dcs_write(dsi, MIPI_DCS_SET_COLUMN_ADDRESS, payload,
                                 sizeof(payload));
        if (err < 0)
                return err;

        return 0;
}
EXPORT_SYMBOL(mipi_dsi_dcs_set_column_address);
```
I don't really use C often.

**Comment by dlech** [comment]
> Is there any documentation on the standard use/parameters of these MIPI commands?

The [datasheet](https://cdn-shop.adafruit.com/datasheets/ST7735R_V0.2.pdf) should explain the commands (just look at the numbers since they don't have the same name).



> but I'm not sure what to modify here.

Hopefully you just need to add a few pixels to `start` and `end`. So `start += 2;` and `end += 2` at the beginning of the function.

**Comment by luRaichu** [comment]
I did
```c
int mipi_dsi_dcs_set_column_address(struct mipi_dsi_device *dsi, u16 start += 1,
                                    u16 end -= 1)
...
```
But there's no effect.

**Comment by dlech** [comment]
You have to add it in the code, not in the function signature. I would expect your change to cause a compile error.

```c
int mipi_dsi_dcs_set_column_address(struct mipi_dsi_device *dsi, u16 start,
                                    u16 end)
{
        const u16 offset = 2; /* change this number as needed */
        start += offset;
        end += offset;
        u8 payload[4] = { start >> 8, start & 0xff, end >> 8, end & 0xff };
        ssize_t err;

        ...
}
```

**Comment by luRaichu** [comment]
> You have to add it in the code, not in the function signature. I would expect your change to cause a compile error.
> 
> ```c
> int mipi_dsi_dcs_set_column_address(struct mipi_dsi_device *dsi, u16 start,
>                                     u16 end)
> {
>         const u16 offset = 2; /* change this number as needed */
>         start += offset;
>         end += offset;
>         u8 payload[4] = { start >> 8, start & 0xff, end >> 8, end & 0xff };
>         ssize_t err;
> 
>         ...
> }
> ```

I did this instead but still no effect... Pretty sure I'm running the kernel with this change.

**Comment by dlech** [comment]
If the first attempt didn't cause a compile error, then I don't think this file is actually being compiled and used.

**Comment by luRaichu** [comment]
What do we do next?

**Comment by dlech** [comment]
I think this is not the right place to make the change in our version of the kernel.  I think it needs to be changed here:

https://github.com/ev3dev/ev3-kernel/blob/b8e704fe9ed45fcc34a6e5658078bfc5a923fb8d/drivers/gpu/drm/tinydrm/mipi-dbi.c#L235-L240

```c
        const u16 col_ofs = 2;
        const u16 row_ofs = 2;
	mipi_dbi_command(mipi, MIPI_DCS_SET_COLUMN_ADDRESS,
			 ((clip.x1 + col_ofs) >> 8) & 0xFF, (clip.x1 + col_ofs) & 0xFF,
			 ((clip.x2 + col_ofs) >> 8) & 0xFF, (clip.x2 + col_ofs - 1) & 0xFF);
	mipi_dbi_command(mipi, MIPI_DCS_SET_PAGE_ADDRESS,
			 ((clip.y1 + row_ofs) >> 8) & 0xFF, (clip.y1 + row_ofs) & 0xFF,
			 ((clip.y2 + row_ofs) >> 8) & 0xFF, (clip.y2 +row_ofs - 1) & 0xFF);
```

**Comment by luRaichu** [comment]
> I think this is not the right place to make the change in our version of the kernel. I think it needs to be changed here:
> 
> https://github.com/ev3dev/ev3-kernel/blob/b8e704fe9ed45fcc34a6e5658078bfc5a923fb8d/drivers/gpu/drm/tinydrm/mipi-dbi.c#L235-L240
> 
> ```c
>         const u16 col_ofs = 2;
>         const u16 row_ofs = 2;
> 	mipi_dbi_command(mipi, MIPI_DCS_SET_COLUMN_ADDRESS,
> 			 ((clip.x1 + col_ofs) >> 8) & 0xFF, (clip.x1 + col_ofs) & 0xFF,
> 			 ((clip.x2 + col_ofs) >> 8) & 0xFF, (clip.x2 + col_ofs - 1) & 0xFF);
> 	mipi_dbi_command(mipi, MIPI_DCS_SET_PAGE_ADDRESS,
> 			 ((clip.y1 + row_ofs) >> 8) & 0xFF, (clip.y1 + row_ofs) & 0xFF,
> 			 ((clip.y2 + row_ofs) >> 8) & 0xFF, (clip.y2 +row_ofs - 1) & 0xFF);
> ```

I just tested this now. No effect.

**Comment by dlech** [comment]
How did you package and install the kernel? What does `uname -a` show?

**Comment by luRaichu** [comment]
The kernel is properly installed.
```
Linux ev3dev 4.14.117-ev3dev-2.3.5-luraichu-fermion-tft-ev3+ #11 PREEMPT Mon Jan 15 17:14:30 UTC 2024 armv5tejl GNU/Linux
```

**Comment by luRaichu** [comment]
Actually, this works! Turns out the release name had changed for the newly built kernel and dpkg refused to flash it. (oh, and `col_ofs` is supposed to be just 1, btw)
Here's my kernel build as a Debian package, with the screen offsets corrected for users of the Fermion 1.8" TFT.
[linux-image-4.14.117-ev3dev-2.3.5-luraichu-fermion-tft-ev3+_1_armel.deb.tar.gz](https://github.com/ev3dev/ev3dev/files/14186438/linux-image-4.14.117-ev3dev-2.3.5-luraichu-fermion-tft-ev3%2B_1_armel.deb.tar.gz)
Unarchive with `tar -xvf linux-image*.deb`, then install with `sudo dpkg -i linux-image*.deb`. If you're already running a custom kernel with an altered release name, you'll need to force install this.
![ev3_brick_offset_fixed](https://github.com/ev3dev/ev3dev/assets/69605426/57f1aedb-4c0a-44ec-b0a8-0c67034dbf10)
